🚀 4단계 - ControllerAdvice, ExceptionHandler 구현하기

djawnstj opened the pull request:
안녕하세요, 정완님!
정성스러운 리뷰 항상 감사합니다ㅠㅠ

이번 4단계는 간단할거라 생각 했는데 미션 하다 보니 큰 문제가 모듈 간 순환 참조가 발생해버리더라고요ㅠㅠㅠ
mvc 와 di 모듈 간 순환 참조 문제가 발생 했는데 순환 참조를 제거하려다 보니 aop 모듈에서 `BeanFactoryBeanPostProcessor` 를 만들고 di 모듈에서 `AnnotationConfigWebApplicationContext` 을 수정하는 방법으로 해결 했습니다.
그러다 보니 코드가 조금 난잡해 보이게 됐네요ㅠㅠ

djawnstj committed:
chore: mvc 모듈에서 aop 모듈 의존

djawnstj committed:
feat: @Configuration 애너테이션에서 @Component 애너테이션을 중첩으로 가지도록 수정

djawnstj committed:
feat: ControllerAdvice 애너테이션 추가

djawnstj committed:
feat: ExceptionHandler 애너테이션 추가

djawnstj committed:
feat: ControllerAdvice 의 메서드를 실행시켜 줄 ExceptionHandlerMethodResolver 개발

djawnstj committed:
feat: ControllerAdvice 의 메서드들에서 처리해 줄 예외와 메서드를 분리시켜 등록 하는 ExceptionHandlerExceptionResolver 개발

djawnstj committed:
feat: 빈 초기화 시점에서 ControllerAdvice 빈을 후처리 하는 ControllerAdviceBeanPostProcessor 개발

djawnstj committed:
feat: HandlerExceptionResolver 인터페이스 개발

djawnstj committed:
feat: DispatcherServlet 에서 HandlerExceptionResolver 를 가지고 있는 HandlerExceptionRegistry 개발

djawnstj committed:
feat: HandlerExceptionResolver 자동 빈 등록

djawnstj committed:
feat: Servlet 예외 처리

djawnstj committed:
test: 테스트를 위한 객체

ghojeong commented on aop/src/main/java/com/interface21/beans/factory/config/BeanFactoryBeanPostProcessor.java:
```diff
@@ -0,0 +1,5 @@
+package com.interface21.beans.factory.config;
+
+public abstract class BeanFactoryBeanPostProcessor implements BeanPostProcessor {
+    public abstract void injectBeanFactory(final Object beanFactory);
+}
```
```suggestion
public interface BeanFactoryBeanPostProcessor extends BeanPostProcessor {
    void injectBeanFactory(final Object beanFactory);
}
```

interface 가 아니라 굳이 abstract class 로 하신 이유가 있으시나요?

djawnstj replied:
앗 정신 없이 작업 하다 보니 추상클래스로 만들어버렸어요ㅠㅠ
인터페이스로 수정하는 것이 더 좋을 것 같습니다.

ghojeong commented on di/src/main/java/com/interface21/context/support/AnnotationConfigWebApplicationContext.java:
```diff
@@ -22,6 +23,20 @@ public class AnnotationConfigWebApplicationContext implements ApplicationContext
     private final DefaultListableBeanFactory beanFactory;
 
     public AnnotationConfigWebApplicationContext(Class<?>... annotatedClasses) {
+        beanFactory = initBeanFactory(annotatedClasses);
+        beanFactory.preInstantiateSingletons();
+    }
+
+    public AnnotationConfigWebApplicationContext(final BeanFactoryBeanPostProcessor beanPostProcessor, final Class<?>... annotatedClasses) {
+        beanFactory = initBeanFactory(annotatedClasses);
+        beanFactory.addBeanPostProcessor(beanPostProcessor);
+        beanPostProcessor.injectBeanFactory(beanFactory);
+
+        beanFactory.preInstantiateSingletons();
+    }
+
+    private DefaultListableBeanFactory initBeanFactory(final Class<?>[] annotatedClasses) {
+        final DefaultListableBeanFactory beanFactory;
```
> 이번 4단계는 간단할거라 생각 했는데 미션 하다 보니 큰 문제가 모듈 간 순환 참조가 발생해버리더라고요ㅠㅠㅠ
mvc 와 di 모듈 간 순환 참조 문제가 발생 했는데 순환 참조를 제거하려다 보니 aop 모듈에서 BeanFactoryBeanPostProcessor 를 만들고 di 모듈에서 AnnotationConfigWebApplicationContext 을 수정하는 방법으로 해결 했습니다.
그러다 보니 코드가 조금 난잡해 보이게 됐네요ㅠㅠ

ㅋㅋㅋㅋ 3단계 때 모듈간 의존성에 관해서 언급할까 하다가,
문제가 반드시 발생할 것 같아서 넘겼는데 역시나 이군요.

코드가 난잡해보인다고 하셨는데, 이 정도면 괜찮은 것 같습니다.
다만 추후에 난잡해질 수 있는 가능성에 대해서 질문을 하나 드리려고 합니다.

app 모듈에서 aop, di, jdbc, mvc 에 대해서 의존성을 가지고 있는 것은 자연스러워보입니다.
하지만 di 에서 aop 와 jdbc 에 반드시 의존적이어야 하는지에 관해서는 다시 한번 생각해볼 가치가 있을 것 같습니다.

지금과 같은 상황에서는 Bean 등록이 필요한 모듈이 생길 때마다,
반드시 di 모듈에서 의존성을 추가해주어야 할 것 같은데,
모듈간의 의존성을 역전시킬 방법이 있을지 질문을 해봅니다.

지금과 같은 상황에서는 db 커넥션과, aop 기능 필요없이,
순수하게 di 기능만을 활용해서 간단한 콘솔 프로그램을 만들려고 해도,
jdbc 와 aop 에 대한 의존성이 생겨버립니다.

사다리 타기 게임처럼 단순한 콘솔 프로그램을 만들 때 싱글톤 디자인 패턴을 적용하기 귀찮아서,
DI 프레임워크의 도입을 하려고 한다면 모듈간의 의존성 설계를 어떻게 하면 좋을까요?

공부를 해보시고 답변 주시면 좋을 것 같습니다.
힌트를 하나 드리자면 과거에 Spring 에서 DI 기능을 활용하려할 때,
왜 XML 이 필요했는지 공부해보시면 좋을 것 같습니다.

djawnstj replied:
생각해보니 di 모듈에서 aop 와 jdbc 에 의존하고 있다는게 어색하긴 하네요.

DI 프레임워크를 사용한다면, 빈 등록을 따로 해주면 해결될 것 같습니다.
Configuration 처럼 등록을 시켜줄 수 있을 것 같습니다!

ghojeong commented on di/src/main/java/com/interface21/beans/factory/support/DefaultListableBeanFactory.java:
```diff
@@ -31,6 +32,8 @@ public class DefaultListableBeanFactory implements BeanDefinitionRegistry, Confi
 
     @Override
     public void preInstantiateSingletons() {
+        registerBean(BeanFactory.class, this);
```
BeanFactory 에서 자기 자신을 Bean 으로 등록하는 군요.
이 부분은 왜 이럴 필요가 있었는지 이해가 어려운데, 설명으로 도와주실 수 있으신가요?
AnnotationConfigWebApplicationContext 에서 initBeanFactory 를 통해 BeanFactory 인스턴스를 생성하고, 멤버로 관리를 하고 있어서요.

BeanFactory 가 자기 자신을 관리를 할 필요성이 생긴 이유가 무엇인가요?

`beanFactory.getBean(BeanFactory.class).getBean(BeanFactory.class)...` 
이렇게 재귀적으로 getBean 을 해올 수가 있는데, 이런 상황을 허용할 필요성이 있는지 여쭈어봅니다.

djawnstj replied:
스프링도 ApplicationContext, BeanFactory 와 같은 클래스를 빈으로 등록 하는 것으로 알고 있습니다.
저는 WebMvcConfigurationSupport 클래스에서 ExceptionHandlerExceptionResolver 를 빈으로 등록했는데, ExceptionHandlerExceptionResolver 가 내부적으로 BeanFactory 를 상태로 가지고 있어야 했습니다.

그 이유가 @ControllerAdvice 로 예외 처리하는 메서드의 파라미터를 생성하기 위함이었는데요, 지금 미션 요구사항으로는 굳이 필요 없는 기능 같긴 하네요!

ghojeong replied:
앗 저도 다시 공부해보니, 등록하고 있다는 부분 확인할 수 있었네요.
빈 등록 후 실제로 사용하는 부분에 대한 코드를 찾기 어려워서 질문드렸는데,
Spring 의 실제 코드를 참조해서 따라하신 거라면 유지하는게 좋을것 같습니다.

ghojeong commented on mvc/src/test/java/com/interface21/web/exception/ExceptionHandlerMethodResolverTest.java:
```diff
@@ -0,0 +1,31 @@
+
+        // then
+        assertThat(actual).isNotNull();
+    }
```
새로 구현한 ControllerAdvice 와 ExceptionResolver 에 대한 꼼꼼한 테스트 좋습니다. 👍

ghojeong commented on mvc/src/test/java/com/interface21/web/code/FakeBeanFactory.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.util.Set;
+
+public class FakeBeanFactory implements BeanFactory {
+    @Override
+    public Set<Class<?>> getBeanClasses() {
+        return Set.of();
+    }
+
+    @Override
+    public <T> T getBean(final Class<T> clazz) {
+        return null;
+    }
+
+    @Override
+    public void clear() {
+
+    }
+
+    @Override
+    public void addBeanPostProcessor(final BeanPostProcessor beanPostProcessor) {
+
+    }
+}
```
어김없이 Fake 객체 직접 만들어주시는 점 좋습니다. 
다만 di 패키지의 FakeBeanFactory 와 어떻게 다른지 처음보는 사람은 무척 혼란스러울 것 같은데,
차이점이 드러나도록 클래스명을 개선해주시거나,
아니면 왜 두개의 다른 FakeBeanFactory 가 필요한지 주석이 있으면 좋을 것 같습니다.

djawnstj replied:
테스트라도 클래스 명을 잘 분리 해야하는데 꼼꼼하게 하지 못했네요ㅠㅠ

ghojeong commented on mvc/src/main/java/com/interface21/web/exception/ControllerAdviceBeanPostProcessor.java:
```diff
@@ -0,0 +1,33 @@
+import com.interface21.beans.factory.config.BeanFactoryBeanPostProcessor;
+import com.interface21.web.bind.annotation.ControllerAdvice;
+
+public class ControllerAdviceBeanPostProcessor extends BeanFactoryBeanPostProcessor {
```
모듈간의 관계와 객체 배치가 혼란스러워서 질문을 드립니다.
TransactionBeanPostProcessor 은 di 패키지에 위치시켰는데,
ControllerAdviceBeanPostProcessor 는 왜 mvc 패키지에 위치시킨건가요?

ControllerAdvice 가 spring-mvc 기능과 왜 밀접한 관련이 있어야하는지 이해가 가지 않습니다.
가령 철저하게 JSON 만을 응답하고 요청받는 API 서버라면 spring-mvc 가 필요없어서 모듈을 gradle 에서 제외할 수도 있는데,
그러면 ControllerAdvice 기능을 사용하지 못하게 되는 건가요?

djawnstj replied:
일단 TransactionBeanPostProcessor 는 jdbc 나 aop 모듈로 이동하는 것이 맞는 것 같습니다.
ControllerAdviceBeanPostProcessor 는 스프링 처럼 구현하려다 보니 resolveException 파라미터로 Request 와 Response 를 받으려고 했어요. 그러다 보니 자연스럽게 mvc 로 이동했습니다.

추가로 제 개인적인 생각은 html 서빙이 아닌 json 서빙을 하는 역할도 web-mvc 에 있다고 생각 합니다!
애초에 DispatcherServlet::doDispatch 내부에서 예외를 잡아 처리하는 로직으로 이해했습니다.

ghojeong replied:
다시 공부해보니 준서님 말이 맞네요.
[ControllerAdvice 는 aop 보다는 web-mvc](https://github.com/esperar/estudy/blob/master/Back-End%2Fspring%2Fcontrolleradvice.md#controlleradvice%EB%8A%94-aop%EB%A1%9C-%EA%B5%AC%ED%98%84%EB%90%98%EC%96%B4-%EC%9E%88%EB%82%98) 와 더 밀접한 관련이 있는게 맞습니다.

반박 감사합니다!
준서님도 공부하셨던 좋은 자료나 공식문서 있으면 README 등에 정리해주시면 감사하겠습니다.

ghojeong reviewed (changes requested):
엄준서님!
벌써 마지막 단계이군요.
모듈간의 의존성과 관계가 이해가 어려운 부분이 있어 질문을 남겨보았습니다.

준서님이 구현해주신 내용에 따르면 ControllerAdvice 기능을 사용하기 위해선 mvc 패키지가 필요한데,
저는 이러한 모듈 의존성이 바람직하지 않아보입니다.

djawnstj committed:
refactor: 모듈 간 의존 관계 정리

di 모듈이 필요 없는 aop, jdbc 등을 의존 하면서 불필요하게 의존 관계가 많아져 모듈 간 의존을 정리했다.
di 가 스프링의 context, beans 와 유사하다고 판단하여 mvc 가 di 를 의존, aop 가 di 를 의존, jdbc 가 di 와 aop 를 의존하도록 변경하였다.

djawnstj committed:
test: 테스트에서 사용 되는 가짜 객체 이름 변경

djawnstj commented:
정완님, 피드백 반영을 일단 했는데, 모듈 간 의존 관계만 정리 해봤습니다.

__BeanFactory 빈 등록__
결국 TransactionBeanPostProcessor 에서 프록시를 만들 때 BeanFactory 가 필요해서 자기 자신을 빈 등록 하는 코드는 유지가 됐어요.
(https://github.com/next-step/java-aop/pull/23#discussion_r1732688864)

__의존 관계 정리__
의존 관계는
di
mvc: di
aop: di
jdbc: di, aop
app: di, mvc, aop, jdbc
이렇게 의존이 되도록 구현 했습니다.

__ControllerAdvice__
사실 저는 혼자 스프링 프레임워크 만들기를 해본 경험이 있어요. 
그때 컨트롤러 예외 처리를 하면서 스프링 코드를 자세히 본 경험이 있습니다.(https://djawnstj.tistory.com/29)
그래서 미션 보자마자' `HandlerExceptionResolver` 인터페이스를 구현 해야 하는구나!' 생각을 했어요ㅋㅋㅋ
AOP를 잘 모르다 보니 Advice 라는 이름을 보고도 AOP/프록시를 먼저 떠올리지 못한 것 같습니다.

스프링 코드에서 DispatcherServlet 을 따라가다 보면
[DispatcherServlet::doDispatch](https://github.com/spring-projects/spring-framework/blob/6d1294ddad256ac3105fdf250051a570006f2dfe/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/ExceptionHandlerExceptionResolver.java#L554) - [ExceptionHandlerExceptionResolver::doResolveHandlerMethodException](https://github.com/spring-projects/spring-framework/blob/6d1294ddad256ac3105fdf250051a570006f2dfe/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/ExceptionHandlerExceptionResolver.java#L554) 에서 `ExeptionHandlerExceptionResolver`-`ExceptionHandlerMethodResolver` 를 호출 하는 것으로 확인 했습니다!

ghojeong reviewed (approved):
엄준서님.
함께 미션을 진행하며 "많이 배웠습니다."
정말 즐거웠고 오랜만에 더 치열하게 개발해야겠다는 자극이 되었습니다.
완주 축하드립니다. 🎉 🎉 🎉 🎉
